### PR TITLE
locale/timezone configs, concurrency issues, specs

### PR DESCRIPTION
**Pull request by colinsurprenant**
moving original PR from elasticsearch/logstash#1279
- also added `timezone` config
- also fixed thread safety issue in shutdown logic
- re-enabled & fixed all specs

### CODE REVIEW

File: lib/logstash/inputs/syslog.rb [hunk 1]
```diff
@@ -59,8 +62,10 @@ def register
       "tag_on_failure" => ["_grokparsefailure_sysloginput"],
     )
 
+    locale = @config["locale"][0] if @config["locale"] != nil and @config["locale"][0] != nil
```
Comment (suyograo) [inline lib/logstash/inputs/syslog.rb]: Hi @nwknu, we should remove this check as @colinsurprenant pointed out and directly pass it to the date filter if it is not nil or empty

File: lib/logstash/inputs/syslog.rb [hunk 2]
```diff
@@ -1,10 +1,11 @@
 # encoding: utf-8
 require "date"
+require "socket"
+require "atomic"
```
Comment (jordansissel) [inline lib/logstash/inputs/syslog.rb]: ⚛

File: lib/logstash/inputs/syslog.rb [hunk 3]
```diff
@@ -61,7 +80,9 @@ def register
     )
 
     @date_filter = LogStash::Filters::Date.new(
-      "match" => [ "timestamp", "MMM  d HH:mm:ss", "MMM dd HH:mm:ss", "ISO8601"]
+      "match" => [ "timestamp", "MMM  d HH:mm:ss", "MMM dd HH:mm:ss", "ISO8601"],
+      "locale" => @locale,
```
Comment (suyograo) [inline lib/logstash/inputs/syslog.rb]: do we need to check for `nil` or unset values? what happens when we send nil to the date filter? does it barf?

Comment (colinsurprenant) [inline lib/logstash/inputs/syslog.rb]: no we don't, the date filter already correctly deals with this.

Comment (colinsurprenant) [inline lib/logstash/inputs/syslog.rb]: plus the specs shows that it works correct with or without the `locale`/`timezone` config

File: lib/logstash/inputs/syslog.rb [hunk 4]
```diff
@@ -91,12 +111,11 @@ def run(output_queue)
       begin
         tcp_listener(output_queue)
       rescue => e
-        break if @shutdown_requested
-        @logger.warn("syslog tcp listener died",
-                     :address => "#{@host}:#{@port}", :exception => e,
-                     :backtrace => e.backtrace)
-        sleep(5)
-        retry
+        unless @shutdown_requested.value
```
Comment (ph) [inline lib/logstash/inputs/syslog.rb]: Can we replace this with a predicate? 
`shutdown_requested?`

File: lib/logstash/inputs/syslog.rb [hunk 5]
```diff
@@ -165,7 +184,7 @@ def tcp_listener(output_queue)
 
   public
   def teardown
-    @shutdown_requested = true
+    @shutdown_requested.update{|v| true}
```
Comment (ph) [inline lib/logstash/inputs/syslog.rb]: styling?

```ruby
@shutdown_requested.update { |v| true }
```

File: logstash-input-syslog.gemspec [hunk 6]
```diff
@@ -22,6 +22,7 @@ Gem::Specification.new do |s|
   # Gem dependencies
   s.add_runtime_dependency 'logstash', '>= 1.4.0', '< 2.0.0'
 
+  s.add_runtime_dependency 'atomic'
```
Comment (ph) [inline logstash-input-syslog.gemspec]: Would it be better to use the `the ruby-concurrency` gem?
Since the atomic gem will or is merged with it?

Readme extract:
```
This library has been moved to https://github.com/ruby-concurrency/atomic.
It will be merged with the concurrent-ruby gem under the same org soon.
```

File: spec/inputs/syslog_spec.rb [hunk 7]
```diff
@@ -1,13 +1,13 @@
-# coding: utf-8
+# encoding: utf-8
 require "logstash/devutils/rspec/spec_helper"
 require "socket"
 require "logstash/inputs/syslog"
 require "logstash/event"
+require "stud/try"
```
Comment (ph) [inline spec/inputs/syslog_spec.rb]: Sort the require order here:
- Logstash file
- other libraries

File: spec/inputs/syslog_spec.rb [hunk 8]
```diff
@@ -68,11 +68,56 @@
...
+      socket.close
+
+      events = event_count.times.collect { queue.pop }
+
+      insist { events.length } == event_count
+      events.each do |event|
+        insist { event["@timestamp"].to_iso8601 } == "#{Time.now.year}-10-26T15:19:25.000Z"
       end
     end
   end
```
Comment (ph) [inline spec/inputs/syslog_spec.rb]: Missing a test if we don't define the `locale` option, not sure how the date parser will work if we send `nil`

File: lib/logstash/inputs/syslog.rb [hunk 9]
```diff
@@ -44,10 +45,28 @@ class LogStash::Inputs::Syslog < LogStash::Inputs::Base
...
+  config :timezone, :validate => :string
+
+  # Specify a locale to be used for date parsing using either IETF-BCP47 or POSIX language tag.
+  # Simple examples are `en`,`en-US` for BCP47 or `en_US` for POSIX.
+  # If not specified, the platform default will be used.
+  #
+  # The locale is mostly necessary to be set for parsing month names (pattern with MMM) and
+  # weekday names (pattern with EEE).
+  #
+  config :locale, :validate => :string
```
Comment (ph) [inline lib/logstash/inputs/syslog.rb]: since the locale is not mandatory, are we assuming we are using the system default?

File: lib/logstash/inputs/syslog.rb [hunk 10]
```diff
@@ -44,10 +45,28 @@ class LogStash::Inputs::Syslog < LogStash::Inputs::Base
...
   config :severity_labels, :validate => :array, :default => [ "Emergency" , "Alert", "Critical", "Error", "Warning", "Notice", "Informational", "Debug" ]
 
+  # Specify a time zone canonical ID to be used for date parsing.
+  # The valid IDs are listed on the [Joda.org available time zones page](http://joda-time.sourceforge.net/timezones.html).
+  # This is useful in case the time zone cannot be extracted from the value,
+  # and is not the platform default.
+  # If this is not specified the platform default will be used.
+  # Canonical ID is good as it takes care of daylight saving time for you
+  # For example, `America/Los_Angeles` or `Europe/France` are valid IDs.
+  config :timezone, :validate => :string
```
Comment (ph) [inline lib/logstash/inputs/syslog.rb]: See my comment on `locale` https://github.com/logstash-plugins/logstash-input-syslog/pull/1/files#r22660947

File: lib/logstash/inputs/syslog.rb (hunk 5 above)
Comment (colinsurprenant) [inline lib/logstash/inputs/syslog.rb]: meh, personally I am not a fan of extra spaces for `{ }`... I prefer without, just like parenthesis.

File: logstash-input-syslog.gemspec (hunk 6 above)
Comment (colinsurprenant) [inline logstash-input-syslog.gemspec]: its not merged but just a dependency AFAIK ?

Comment (colinsurprenant) [inline logstash-input-syslog.gemspec]: NOT https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent/atomic.rb

I guess I will update.

File: lib/logstash/inputs/syslog.rb (hunk 5 above)
Comment (ph) [inline lib/logstash/inputs/syslog.rb]: at least add the space after the update

File: spec/inputs/syslog_spec.rb (hunk 7 above)
Comment (colinsurprenant) [inline spec/inputs/syslog_spec.rb]: +1

File: spec/inputs/syslog_spec.rb (hunk 8 above)
Comment (colinsurprenant) [inline spec/inputs/syslog_spec.rb]: most of the tests do not set a `locale` option... not sure what you mean?

File: lib/logstash/inputs/syslog.rb (hunk 9 above)
Comment (colinsurprenant) [inline lib/logstash/inputs/syslog.rb]: "# If not specified, the platform default will be used."  ???

**Comment by ph** [comment]
Tests are running for me, made a few comments.
I don't see any tests related to the addition of `Atomic.new` ? Do we have a ticket for this?

File: spec/inputs/syslog_spec.rb (hunk 8 above)
Comment (ph) [inline spec/inputs/syslog_spec.rb]: -1 on me, I reread the test, I was looking for something more explicit.

File: lib/logstash/inputs/syslog.rb (hunk 10 above)
Comment (colinsurprenant) [inline lib/logstash/inputs/syslog.rb]: " # If this is not specified the platform default will be used." ?

File: lib/logstash/inputs/syslog.rb (hunk 9 above)
Comment (ph) [inline lib/logstash/inputs/syslog.rb]: -1 @ph + 1 @colinsurprenant

File: spec/inputs/syslog_spec.rb (hunk 8 above)
Comment (colinsurprenant) [inline spec/inputs/syslog_spec.rb]: well, actually, I guess this is still a valid suggestion to add an explicit "no options" test.

**Comment by colinsurprenant** [comment]
there was not ticket for the threadsafety stuff using Atomic - just my concurrency eagle eye while going through that code.

File: lib/logstash/inputs/syslog.rb (hunk 5 above)
Comment (colinsurprenant) [inline lib/logstash/inputs/syslog.rb]: why? when you call a method:
```ruby
do_something(with_a_parameter)
```
we don't add extra spaces?

what is the difference? (I am being a dick here I know, but this is actually a styling issue I am struggling with - please help me find peace & balance with it ;)

Comment (ph) [inline lib/logstash/inputs/syslog.rb]: I found the second and the last one easier to read.
I have looked in our STYLE.md and we don't anything for that.

```ruby
@shutdown_requested.update{|v| true}
@shutdown_requested.update {|v| true}
@shutdown_requested.update { |v| true }
```
I could be picky and check for 80 characters line :boom:

File: spec/inputs/syslog_spec.rb (hunk 8 above)
Comment (suyograo) [inline spec/inputs/syslog_spec.rb]: +1

File: lib/logstash/inputs/syslog.rb (hunk 5 above)
Comment (colinsurprenant) [inline lib/logstash/inputs/syslog.rb]: :rocket:

**Comment by colinsurprenant** [comment]
addressed previous comments and found and fixed problem found by @jsvd with failing tests that hanged the pipeline.

File: lib/logstash/inputs/syslog.rb [hunk 11]
```diff
@@ -61,43 +80,25 @@ def register
...
+      "match" => [ "timestamp", "MMM  d HH:mm:ss", "MMM dd HH:mm:ss", "ISO8601"],
+      "locale" => @locale,
+      "timezone" => @timezone,
     )
 
     @grok_filter.register
     @date_filter.register
 
-    @tcp_clients = ThreadSafe::Array.new
+    @tcp_sockets = ThreadSafe::Array.new
```
Comment (jordansissel) [inline lib/logstash/inputs/syslog.rb]: Days since last concurrency violation: 0

<3

File: lib/logstash/inputs/syslog.rb [hunk 12]
```diff
@@ -133,39 +153,46 @@ def udp_listener(output_queue)
 
   private
   def tcp_listener(output_queue)
+    @tcp_sockets = []
     @logger.info("Starting syslog tcp listener", :address => "#{@host}:#{@port}")
     @tcp = TCPServer.new(@host, @port)
-    @tcp_clients = []
```
Comment (colinsurprenant) [inline lib/logstash/inputs/syslog.rb]: OMG this is nasty - `@tcp_clients` / `@tcp_sockets` is reinitialized as a non thread-safe Array here - it was previously initialized as a `ThreadSafe::Array` in `register`.

**Comment by colinsurprenant** [comment]
... updated PR title, as I completely blew away the original PR scope, sorry :P

File: lib/logstash/inputs/syslog.rb [hunk 13]
```diff
@@ -107,13 +108,32 @@ def run(output_queue)
...
+  private
+  def tcp_server(output_queue)
+    tcp_listener(output_queue)
+  rescue => e
+    if @shutdown_requested.false?
+      @logger.warn("syslog tcp listener died", :address => "#{@host}:#{@port}", :exception => e, :backtrace => e.backtrace)
+      sleep(5)
+      retry
     end
+  end
```
Comment (ph) [inline lib/logstash/inputs/syslog.rb]: The `udp_server` and the `tcp_server` method are really similar, we could refactor them into 1 method?

File: lib/logstash/inputs/syslog.rb [hunk 14]
```diff
@@ -135,37 +155,43 @@ def udp_listener(output_queue)
...
+
+    LogStash::Util::set_thread_name("input|syslog|tcp|#{ip}:#{port}}")
+
+    socket.each do |line|
+      @codec.decode(line) do |event|
+        decorate(event)
+        event["host"] = ip
+        syslog_relay(event)
+        output_queue << event
+      end
```
Comment (ph) [inline lib/logstash/inputs/syslog.rb]: `@codec.decode()` in the tcp receiver and the udp receiver implement the same logic, the only different part is the source of the IP address we could refactor this into one method.

File: lib/logstash/inputs/syslog.rb (hunk 13 above)
Comment (ph) [inline lib/logstash/inputs/syslog.rb]: Those two methods rescue all the errors from the sockets, we should be more specific with our errors handling. Some of the errors could come from the `codecs` and could be fatal error?

File: lib/logstash/inputs/syslog.rb [hunk 15]
```diff
@@ -135,37 +155,43 @@ def udp_listener(output_queue)
...
-    end # loop do
+      socket = @tcp.accept
+      @tcp_sockets << socket
+
+      break if @shutdown_requested.true?
+
+      Thread.new(output_queue, socket) do |output_queue, socket|
+        tcp_receiver(output_queue, socket)
+      end
+    end
```
Comment (ph) [inline lib/logstash/inputs/syslog.rb]: We are creating new threads on every new connection and we don't keep track of them. What are the best practice for that? Should we keep a reference in a list to close them on teardown?

File: spec/inputs/syslog_spec.rb [hunk 16]
```diff
@@ -66,13 +72,103 @@
...
+      }
+    CONFIG
+
+    input do |pipeline, queue|
+      t = Thread.new { pipeline.run }
+      sleep 0.1 while !pipeline.ready?
+
+      socket = Stud.try(5.times) { TCPSocket.new("127.0.0.1", port) }
+      socket.puts("<164>Oct 26 15:19:25 1.2.3.4 %ASA-4-106023: Deny udp src DRAC:10.1.2.3/43434 dst outside:192.168.0.1/53 by access-group \"acl_drac\" [0x0, 0x0]")
+      socket.close
```
Comment (ph) [inline spec/inputs/syslog_spec.rb]: I saw this used twice, we could move this into a `let`
```
"<164>Oct 26 15:19:25 1.2.3.4 %ASA-4-106023: Deny udp src DRAC:10.1.2.3/43434 dst outside:192.168.0.1/53 by access-group \"acl_drac\" [0x0, 0x0]"
```
other reference: https://github.com/logstash-plugins/logstash-input-syslog/pull/1/files#diff-6bffb5d7bae5d75fb475af76a37a54eaR109

File: lib/logstash/inputs/syslog.rb (hunk 13 above)
Comment (colinsurprenant) [inline lib/logstash/inputs/syslog.rb]: agree, I will refactor both decode blocks and wrap decoding exception so that they are excluded in the upper server rescue + retry and the retry logic only happen on socket exceptions.

File: lib/logstash/inputs/syslog.rb (hunk 14 above)
Comment (colinsurprenant) [inline lib/logstash/inputs/syslog.rb]: +1

File: lib/logstash/inputs/syslog.rb (hunk 15 above)
Comment (colinsurprenant) [inline lib/logstash/inputs/syslog.rb]: in this case I think this is pretty safe, each thread is associated with a socket which is actually tracked in the @tcp_sockets array and any socket exception (which will happen if we force close the socket) will terminate the Thread. I am not sure that tracking the threads in a collection would provide anything more? 

I think the only thing we could do here is cap the number of open sockets? what do we do when that happen? stop accepting connections until we are below that cap? as it is it could be exploited as a ddos to create an unlimited amount of threads or at least until socket exhaustion... 

thoughts?

Comment (ph) [inline lib/logstash/inputs/syslog.rb]: Some sort of threadpool? 
There is some primitive available in concurrent-ruby
http://ruby-concurrency.github.io/concurrent-ruby/Concurrent/FixedThreadPool.html
and http://ruby-concurrency.github.io/concurrent-ruby/Concurrent/CachedThreadPool.html

This could be done in another PR, I don't think we should complexify this "locales" ticket ;)

Comment (colinsurprenant) [inline lib/logstash/inputs/syslog.rb]: yeah... we could go the threadpool route, but blocking on the max number of open sockets would achieve the same result. agree, let's keep that for another PR :P

File: spec/inputs/syslog_spec.rb (hunk 16 above)
Comment (ph) [inline spec/inputs/syslog_spec.rb]: Same comment, would you mind using a let for this?

File: lib/logstash/inputs/syslog.rb [hunk 17]
```diff
@@ -107,65 +109,94 @@ def run(output_queue)
...
+  # any listener uncatched exception
+  #
+  # @param protocol [Symbol] either :udp or :tcp
+  # @param output_queue [Queue] the pipeline input to filters queue
+  def server(protocol, output_queue)
+    self.send("#{protocol.to_s}_listener", output_queue)
+  rescue => e
+    if @shutdown_requested.false?
+      @logger.warn("syslog #{protocol.to_s} listener died", :address => "#{@host}:#{@port}", :exception => e, :backtrace => e.backtrace)
+      sleep(5)
```
Comment (ph) [inline lib/logstash/inputs/syslog.rb]: We don't really need string interpolation here?
```ruby
:protocol => procotol
``

File: lib/logstash/inputs/syslog.rb [hunk 18]
```diff
@@ -107,65 +109,94 @@ def run(output_queue)
   end # def run
 
   private
+  # server call the specified protocol listener and basically restarts on
+  # any listener uncatched exception
+  #
+  # @param protocol [Symbol] either :udp or :tcp
+  # @param output_queue [Queue] the pipeline input to filters queue
+  def server(protocol, output_queue)
+    self.send("#{protocol.to_s}_listener", output_queue)
```
Comment (ph) [inline lib/logstash/inputs/syslog.rb]: symbols will do an implicite conversion to string when used in string interpolation.

**Comment by ph** [comment]
made minor comments, other than that LGTM.

**Comment by colinsurprenant** [comment]
:+1:

**Comment by ph** [comment]
LGTM

**Comment by colinsurprenant** [comment]
after a few tries :P squashed my commit and kept @nwknu original commits for proper attribution!

File: lib/logstash/inputs/syslog.rb (hunk 15 above)
Comment (colinsurprenant) [inline lib/logstash/inputs/syslog.rb]: created issue #2 for this.

**Comment by colinsurprenant** [comment]
this is included in version 0.1.2
